modin/engines/base/frame/data.py

```diff
@@ -1746,26 +1733,38 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
 
         index_other_obj = [o.axes[axis] for o in other]
         joined_index = self._join_index_objects(axis, index_other_obj, how, sort)
+
         # We have to set these because otherwise when we perform the functions it may
         # end up serializing this entire object.
         left_old_idx = self.axes[axis]
         right_old_idxes = index_other_obj
 
-        def make_map_func():
-            if not joined_index.is_unique and axis == 0:
+        def make_map_func(index, left=True):
+            # left - specific argument for case of binary operation;
+            # it choose indexer for left or right index
+            if index.equals(joined_index):
                 return lambda df: df
+            if reindexer == "binary":
```

Comment:
We should have logic about binary reindexing in a function, that responsible for doing binary operations. Also approach with `reindexer == "something"` don't scale well, imagine huge if-elif-else construction when we would add more cases for reindexing, maybe we can pass `reindexer` like callable or something like that?